agent/pipeline_parser.go

```diff
@@ -14,80 +15,95 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+// PipelineParser parses a pipeline, optionally interpolating values from
+// a given environment.
 type PipelineParser struct {
 	Env             env.Environment
 	Filename        string
 	Pipeline        []byte
 	NoInterpolation bool
 }
 
-func (p PipelineParser) Parse() (*PipelineParserResult, error) {
+// Parse runs the parser.
+func (p *PipelineParser) Parse() (*PipelineParserResult, error) {
 	if p.Env == nil {
 		p.Env = env.New()
 	}
 
-	var errPrefix string
-	if p.Filename == "" {
-		errPrefix = "Failed to parse pipeline"
-	} else {
+	errPrefix := "Failed to parse pipeline"
+	if p.Filename != "" {
 		errPrefix = fmt.Sprintf("Failed to parse %s", p.Filename)
 	}
 
-	var pipelineAsSlice []topLevelStep
-	var pipeline *yaml.Node
+	// Run the pipeline through the YAML parser. Parse to a *yaml.Node because
+	// the format is flexible.
+	var doc yaml.Node
+	if err := yaml.Unmarshal(p.Pipeline, &doc); err != nil {
+		return nil, fmt.Errorf("%s: %v", errPrefix, formatYAMLError(err))
+	}
 
-	// We support top-level arrays of steps, so try that first
-	if err := yaml.Unmarshal(p.Pipeline, &pipelineAsSlice); err == nil {
-		var steps []any
+	// Some quick paranoia checks...
+	if doc.Kind != yaml.DocumentNode {
+		return nil, fmt.Errorf("%s: line %d, col %d: pipeline is not a YAML document? (kind = %x)", errPrefix, doc.Line, doc.Column, doc.Kind)
+	}
+	if len(doc.Content) != 1 {
```

Comment:
I'll add some comments to clarify.

When parsing a document, yaml.v3 creates a DocumentNode and then calls parseChild... only once: https://github.com/go-yaml/yaml/blob/f6f7691b1fdeb513f56608cd2c32c51f8194bf51/decode.go#LL196-L212C2
The child node goes into Content. So doc.Content should contain one item, and that item is the "real" content.

Maybe it's a bug that yaml.v3 only parses one child of the document? Because I don't see where in the YAML spec it's forbidden.

Reply:
Right, gotcha, the `DocumentNode` kind is the root of the yaml tree (for that document, of which there can be multiple[?]), that makes sense.

how would you give a document multiple children?

Reply:
> how would you give a document multiple children?

I imagine something like the following (though it doesn't seem sensible), because the sequence and mapping are different kinds at the same level, they'd have to be separate children of the document:

```yaml
---
- foo
- bar

baz: qux
beep: boop
```

I'm not too sure what yaml.v2 does exactly